lib/ts/recipe/multifactorauth/recipeImplementation.ts

```diff
@@ -1,34 +1,176 @@
+                userContext,
+            });
+
+            const nextFactors = MultiFactorAuthClaim.buildNextArray(completedFactors, mfaRequirementsForAuth);
+            return nextFactors.length === 0 || nextFactors.includes(factorId);
+        },
+
+        markFactorAsCompleteInSession: async function ({ session, factorId, userContext }) {
+            const currentValue = await session.getClaimValue(MultiFactorAuthClaim);
+            const completed = {
+                ...currentValue?.c,
+                [factorId]: Math.floor(Date.now() / 1000),
+            };
+            const setupUserFactors = await this.recipeInterfaceImpl.getFactorsSetupForUser({
+                userId: session.getUserId(),
+                tenantId: session.getTenantId(),
+                userContext,
+            });
+            const requirements = await this.config.getMFARequirementsForAuth(
+                session,
+                setupUserFactors,
+                completed,
+                userContext
+            );
+            const next = MultiFactorAuthClaim.buildNextArray(completed, requirements);
+            await session.setClaimValue(MultiFactorAuthClaim, {
+                c: completed,
+                n: next,
+            });
+        },
+
+        addToDefaultRequiredFactorsForUser: async function ({ tenantId, user, factorId, userContext }) {
+            const userMetadataInstance = UserMetadataRecipe.getInstanceOrThrowError();
+            const metadata = await userMetadataInstance.recipeInterfaceImpl.getUserMetadata({
+                userId: user.id,
+                userContext,
+            });
+
+            const factorIds = metadata.metadata._supertokens?.defaultRequiredFactorIdsForUser?.[tenantId] ?? [];
+            if (factorIds.includes(factorId)) {
+                return;
+            }
+
+            factorIds.push(factorId);
+
+            const metadataUpdate = {
+                ...metadata.metadata,
+                _supertokens: {
+                    ...metadata.metadata._supertokens,
+                    defaultRequiredFactorIdsForUser: {
+                        ...metadata.metadata._supertokens?.factors,
+                        [tenantId]: factorIds,
+                    },
+                },
+            };
+
+            await userMetadataInstance.recipeInterfaceImpl.updateUserMetadataInternal({
+                userId: user.id,
+                metadataUpdate,
+                userContext,
+            });
+        },
+
+        getDefaultRequiredFactorsForUser: async function ({ tenantId, user, userContext }) {
+            const userMetadataInstance = UserMetadataRecipe.getInstanceOrThrowError();
+            const metadata = await userMetadataInstance.recipeInterfaceImpl.getUserMetadata({
+                userId: user.id,
+                userContext,
+            });
+
+            return metadata.metadata._supertokens?.defaultRequiredFactorIdsForUser?.[tenantId] ?? [];
```

Comment:
What happens if a user is shared with a new tenant after setting up MFA? they have to set up all their factors once again?